pkg/cri/server/image_pull.go

```diff
@@ -93,6 +94,19 @@ import (
 
 // PullImage pulls an image with authentication config.
 func (c *criService) PullImage(ctx context.Context, r *runtime.PullImageRequest) (*runtime.PullImageResponse, error) {
+	inProgressImagePulls.Inc()
+	defer inProgressImagePulls.Dec()
+	var domain string
+	var err error
```

Comment:
I guess a common practice of handling error in defer is to use named return value, instead of using a var like here, because in this case you can't be sure `err` will always be the return value. 

For example, if you look at https://github.com/containerd/containerd/blob/b8de211b44721d053e6e597292b1dbd791f02c7a/pkg/cri/server/image_pull.go#L196

the returned error is based on a local `err` in the for loop. So your defer here will not catch that local error. 

(It's similar to https://go.dev/play/p/LNkan9NkAUt)

Reply:
If so, I have to wrap like below.
```
func (c *criService) PullImage(ctx context.Context, r *runtime.PullImageRequest) (*runtime.PullImageResponse, error) {
	resp, domain, err := c.PullImageImp(ctx, r)
	if err != nil {
		imagePulls.WithValues("failed", domain).Inc()
	} else {
		imagePulls.WithValues("succeed", domain).Inc()
	}
	return resp, err
}
```
PullImageImp is the original function. Is this OK?

Reply:
no you dont need to. Just use a named return value for the returned error and use the named return value in your defer, and golang will make sure it is always the returned error. Like this:

```
func (c *criService) PullImage(ctx context.Context, r *runtime.PullImageRequest) (_ *runtime.PullImageResponse, retErr error) {
	defer func() {
		if retErr != nil {
			...
		} else {
			...
		}
	}()
```

Reply:
As I need the domain as a label of the metrics, and the image registry domain is not returned.

This still does not work. 

So I kept the change, would you take a look again?

Reply:
Sorry for the late reply. I am just trying to minimize the number of wrapper methods. Can we do something like

```
func (c *criService) PullImage(ctx context.Context, r *runtime.PullImageRequest) (_ *runtime.PullImageResponse, retErr error) {
        var domain string
	defer func() {
		if retErr != nil {
			...
		} else {
			...
		}
	}()

```

and `domain` will be empty if it errs when trying to parse it, and will be non-empty if the parsing succeeds.